src/NUnitFramework/framework/Internal/Commands/TestMethodCommand.cs

```diff
@@ -79,14 +79,7 @@ private object RunTestMethod(TestExecutionContext context)
 #if ASYNC
             if (AsyncToSyncAdapter.IsAsyncOperation(testMethod.Method.MethodInfo))
             {
-                try
-                {
-                    return AsyncToSyncAdapter.Await(() => InvokeTestMethod(context));
-                }
-                catch (Exception e)
-                {
-                    throw new NUnitException("Rethrown", e);
```

Comment:
I'd like to deprecate `NUnitException` entirely in the future.

Reply:
So not wrapping the exception in a `NUnitException` solves the problem?
(Actually, I got most information from the title of the second commit `Fix twice-wrapped NUnitException to fix Assert.Pass, Ignore, and Inconclusive in non-async Task-returning methods` 😄).

Reply:
NUnitException is an implementation detail that's never supposed to escape from the framework, so there shouldn't be anything to deprecate.

It was used internally to indicate that the wrapped exception has special meaning for NUnit and shouldn't be handled as an exception. It's predecessor in V2 was used more heavily than it is today so you could phase it out of the implementation, but there doesn't seem to be a lot of benefit in putting effort into removing it.

Reply:
@mikkelbu Correct. It was an NUnitException with an InnerException of an NUnitException with the InnerException of the Success/Ignore/InconclusiveException. `ValidateAndUnwrap` only unwraps one level, so `if (ex is ResultStateException)` fails because the ResultStateException is actually `ex.InnerException` at that point.

https://github.com/nunit/nunit/blob/b5c1d549938476ff06f118557b29778b36593549/src/NUnitFramework/framework/Internal/Results/TestResult.cs#L578-L585

https://github.com/nunit/nunit/blob/b5c1d549938476ff06f118557b29778b36593549/src/NUnitFramework/framework/Internal/Results/TestResult.cs#L560-L568

The problem was caused in https://github.com/nunit/nunit/commit/b83f556cc41b3c12a6e1a7729d1e05e5e140f252#diff-6325ab2a4b2ce8d3a6fd52a0f58f75f5L98 by me keeping the `catch` block but failing to notice that it used to only apply to the `WaitForPendingOperationsToComplete` part and not the `Reflect.InvokeMethod` part which already wraps.

Reply:
@mikkelbu I moved the explanation to the PR body. Is there anything else I can clarify?